File: blind 75/Arrays and Strings/longest_palindromic_substring.py
'''
Given a string s, return the longest palindromic substring in s.

 

Example 1:

Input: s = "babad"
Output: "bab"
Note: "aba" is also a valid answer.
Example 2:

Input: s = "cbbd"
Output: "bb"
Example 3:

Input: s = "a"
Output: "a"
Example 4:

Input: s = "ac"
Output: "a"
'''

"""
Revisited on 8/9/2023
Similar solution as below
Time complexity: O(N^2), for each ith character, we evaluate all other characters in the string
Space complexity: O(N), it's possible that the palindrome is the entire length of the original string
"""
class Solution2:
    def longestPalindrome(self, s: str) -> str:
        maxLength = float("-inf")
        substring = ""
        for i in range(len(s)):
            # check for odd length palindrome,
            # where l = i - 1, one element to the left
            # and r = i + 1, one element to the right,
            # and then move both pointers outwards
            l = i-1
            r = i+1
            while True:
                if l < 0 or r > len(s) - 1 or s[l] != s[r]:
                    # length of substring is the (right index - 1) - (left index + 1) + 1
                    if r-l+1> maxLength:
                        maxLength = r-l+1
                        # because we extended one past the boundary, we need to add
                        # one back to l
                        substring = s[l+1:r]
                    break
                else:
                    l -= 1
                    r += 1     
            # check for even length palindrome, where l = i, and r would be one element to the right
            # and move both pointers outwards
            l = i
            r = i+1
            while True:
                if l < 0 or r > len(s) - 1 or s[l] != s[r]:
                    if r-l+1 > maxLength:
                        maxLength = r-l+1
                        substring = s[l+1:r]
                    break
                else:
                    l -= 1
                    r += 1
        
        return substring
            
def longestPalindrome(self, s: str) -> str:
        def longestAtIndex(s, left, right):
            ## starting from the left and right pointers, expand outwards <-- -->
            ## if at any point, the characters are not equal or you're out of bounds 
            ## of the string, break out of the loop
            while (left >= 0 and right < len(s) and s[left] == s[right]):
                left -= 1
                right += 1
            left += 1
            right -= 1
            length = right - left + 1
            return (left, right, length)
        left = 0
        right = -1
        longest = 0
        substr = ''
        for i in range(len(s)):
            ## start the left and right pointers at the same index for odd number palindromes
            odd_l, odd_r, odd_length = longestAtIndex(s, i, i)
            if (odd_length > longest):
                longest = odd_length
                left = odd_l
                right = odd_r
            ## start the left and right pointers at i, i + 1 for even number palindromes
            even_l, even_r, even_length = longestAtIndex(s, i, i + 1)
            if (even_length > longest):
                longest = even_length
                left = even_l
                right = even_r
        ## return the substring
        return s[left: right+1]

File: node_js_ds_and_algos_implementations/1D dynamic programming/longest_palindromic_substring.js
class Solution {
    /**
     * @param {string} s
     * @return {string}
     */
    longestPalindrome(s) {
        // revisited 5/20/2026
        // for each character, set two pointers starting at given index
        // and move them outwards
        // for even length palindromes, start l = i, r = i + 1
        // for odd numbered, start l = i, r = i
        let curLength = 0
        let res = ""
        for (let i = 0; i < s.length; ++i){
            // find odd length
            let l1 = i
            let r1 = i
            while (l1 >= 0 && r1 < s.length && s[l1] === s[r1]){
                --l1
                ++r1
            }
            let l2 = i
            let r2 = i+1
            while (l2 >= 0 && r2 < s.length && s[l2] === s[r2]){
                --l2
                ++r2
            }
            let oddLength = r1-l1-1
            let evenLength = r2-l2-1
            if (oddLength > evenLength && oddLength > curLength){
                res = s.slice(l1+1,r1)
                curLength = oddLength
            }
            else if (oddLength < evenLength && evenLength > curLength) {
                res = s.slice(l2+1, r2)
                curLength = evenLength
            }
        }
        return res
    }
}

/**
 * @param {string} s
 * @return {string}
 */
/* 
Similar approach to length of longest palindromic substring, 
except this time you keep track of the left and right boundaries
once you've found a substring length that exceeds the max.

In the while loop, when exceeding the boundary, you have to shrink the boundary
back by 1 (adding one to the left, subtracting one from the right) since it would
"one too many" due to the structure of the loop

Time Complexity:
O(N^2)
Space:
O(1)
*/
var longestPalindrome = function(s) {
    let N = s.length
    if (N === 1){
        return s[0]
    }
    let maxCount = 0
    let maxIndices = []
    for (let i = 0; i < N; ++i){
        let l1 = i-1
        let r1 = i+1
        let oddCount = 1
        while (true){
            // if we extend past the boundaries, we need to add one back to the left
            // and subtract one back to the right to get the correct boundaries
            if (!(l1 >= 0 && r1 <= N - 1)){
                ++l1
                --r1
                break
            }
            // if we extend past a valid palindromic substring, we need to add one back to the left
            // and subtract one back to the right to get the correct boundaries
            if (s[l1] !== s[r1]){
                ++l1
                --r1
                break
            }
            oddCount += 2
            --l1
            ++r1
        }
        let l2 = i
        let r2 = i + 1
        let evenCount = 0
        while (true){
            if (!(l2 >= 0 && r2 <= N - 1)){
                ++l2
                --r2
                break
            }
            if (s[l2] !== s[r2]){
                ++l2
                --r2
                break    
            }
            evenCount += 2
            --l2
            ++r2
        }
        let indices = oddCount > evenCount ? [l1, r1] : [l2, r2]
        if (oddCount > maxCount){
            maxCount = oddCount
            maxIndices = indices
        }
        else if (evenCount > maxCount){
            maxCount = evenCount
            maxIndices = indices
        }
    }
    return s.slice(maxIndices[0], maxIndices[1]+1)
};